src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -45,127 +53,135 @@ namespace Microsoft.VisualStudio.LanguageServices.Packaging
             SVsServiceProvider serviceProvider,
+            [Import("Microsoft.VisualStudio.Shell.Interop.SAsyncServiceProvider")] object asyncServiceProvider,
             IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
-            [Import(AllowDefault = true)] Lazy<IVsPackageInstallerServices> packageInstallerServices,
-            [Import(AllowDefault = true)] Lazy<IVsPackageInstaller2> packageInstaller,
-            [Import(AllowDefault = true)] Lazy<IVsPackageUninstaller> packageUninstaller,
-            [Import(AllowDefault = true)] Lazy<IVsPackageSourceProvider> packageSourceProvider)
-            : base(threadingContext, workspace, SymbolSearchOptions.Enabled,
-                              SymbolSearchOptions.SuggestForTypesInReferenceAssemblies,
-                              SymbolSearchOptions.SuggestForTypesInNuGetPackages)
+            [Import(AllowDefault = true)] Lazy<IVsPackageInstallerServices>? packageInstallerServices,
+            [Import(AllowDefault = true)] Lazy<IVsPackageInstaller2>? packageInstaller,
+            [Import(AllowDefault = true)] Lazy<IVsPackageUninstaller>? packageUninstaller,
+            [Import(AllowDefault = true)] Lazy<IVsPackageSourceProvider>? packageSourceProvider)
+            : base(threadingContext,
+                   workspace,
+                   SymbolSearchOptions.Enabled,
+                   SymbolSearchOptions.SuggestForTypesInReferenceAssemblies,
+                   SymbolSearchOptions.SuggestForTypesInNuGetPackages)
         {
             _workspace = workspace;
+
             _serviceProvider = serviceProvider;
+            // MEFv2 doesn't support type based contract for Import above and for this particular contract
+            // (SAsyncServiceProvider) actual type cast doesn't work. (https://github.com/microsoft/vs-mef/issues/138)
+            // workaround by getting the service as object and cast to actual interface
+            _asyncServiceProvider = (Shell.IAsyncServiceProvider)asyncServiceProvider;
+
             _editorAdaptersFactoryService = editorAdaptersFactoryService;
             _packageInstallerServices = packageInstallerServices;
             _packageInstaller = packageInstaller;
             _packageUninstaller = packageUninstaller;
             _packageSourceProvider = packageSourceProvider;
+
+            _workQueue = new AsyncBatchingWorkQueue<(bool solutionChanged, ProjectId? changedProject)>(
+                TimeSpan.FromSeconds(1),
+                this.ProcessWorkQueueAsync,
+                equalityComparer: null,
+                listenerProvider.GetListener(FeatureAttribute.PackageInstaller),
+                this.DisposalToken);
         }
 
-        public async ValueTask<ImmutableArray<PackageSource>?> TryGetPackageSourcesAsync(bool allowSwitchToMainThread, CancellationToken cancellationToken)
+        public event EventHandler? PackageSourcesChanged;
+
+        public ImmutableArray<PackageSource> TryGetPackageSources()
```

Comment:
📝 This seems like a big step backwards. The previous code had clearly-articulated failure paths communicated in the return values. The new code limits our ability to address future needs in this code or understand how/when failures are occurring. I would like to see the `bool allowSwitchToMainThread` returned, and the pull request restricted to the change related to the asynchronous APIs.

Reply:
It is not a step backwards.  You added that speculative functionality despite no need for it from the fixed set of use cases we have here.  

> The new code limits our ability to address future needs in this code

No it doesn't.  if we have needs in the future we can add them.

> or understand how/when failures are occurring.

I don't know what failures you're referring to.

>  I would like to see the bool allowSwitchToMainThread returned

I will not be adding that speculative functionality until there is a point that some calling code actually needs that.  

> and the pull request restricted to the change related to the asynchronous APIs.

No.  I think we should take this PR in this form.  It matches the original shape/design we had here, which was created to serve the single use case the IDE ships with (and only intends to ship with for the forseeable future).  At one point you updated it to have this extra functionality, but we've never ever made use of that.  I brought this up as an objection at the time, but didn't feel very strongly against it at the time.  I've since grown to feel that that change was not necessary and was actually a net negative for me in terms of maintenance of this code.  

We should move forward with that parameter removed.  If we need it in the future, we can add it.